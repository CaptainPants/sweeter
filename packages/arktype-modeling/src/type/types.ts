import { Type } from 'arktype';
import { ObjectType } from 'arktype/internal/methods/object.ts';

export type AnyTypeConstraint = Type<any, any>;
export type UnknownType = Type<unknown>;

export type AnyObjectTypeConstraint = ObjectType<object, any>;
export type UnknownObjectType = ObjectType<object, any>;
