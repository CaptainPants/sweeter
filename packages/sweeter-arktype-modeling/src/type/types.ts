import { type Type } from 'arktype';
import { type ObjectType } from 'arktype/internal/methods/object.ts';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyTypeConstraint = Type<any, any>;
export type UnknownType = Type<unknown>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyObjectTypeConstraint = ObjectType<object, any>;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type UnknownObjectType = ObjectType<object, any>;

export interface PropertyInfo<TSchema = Type<unknown>> {
    type: TSchema;
    optional: boolean;
}