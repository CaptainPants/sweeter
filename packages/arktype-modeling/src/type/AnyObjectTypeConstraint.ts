import { ObjectType } from 'arktype/internal/methods/object.ts';

export type AnyObjectTypeConstraint = ObjectType<object, any>;
