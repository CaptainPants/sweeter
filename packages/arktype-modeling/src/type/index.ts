export * from './types.js';
export * from './introspect/index.js';
