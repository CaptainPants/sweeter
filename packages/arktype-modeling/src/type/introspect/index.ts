export * as introspect from './barrel.js';
