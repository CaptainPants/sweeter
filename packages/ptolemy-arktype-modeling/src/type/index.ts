export * from './introspect/index.js';
export * from './types.js';
