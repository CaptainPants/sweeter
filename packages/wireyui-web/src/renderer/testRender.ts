import { WebRenderer } from './WebRenderer.js';

export interface TestRenderResult {
    readonly nodes: NodeListOf<ChildNode>;
    readonly getHTML: () => string;
    readonly dispose: () => void;
}

/**
 * Convenience function for rendering isolated JSX chunks for unit tests
 * @param callback
 * @returns
 */
export function testRender(callback: () => JSX.Element): TestRenderResult {
    const renderer = new WebRenderer();

    const root = document.createElement('div');

    const dispose = renderer.start(root, callback);

    return {
        nodes: root.childNodes,
        getHTML: () => root.innerHTML,
        dispose,
    };
}
