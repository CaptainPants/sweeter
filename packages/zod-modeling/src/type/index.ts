export * from './is/index.js';
export * from './unwrap.js';