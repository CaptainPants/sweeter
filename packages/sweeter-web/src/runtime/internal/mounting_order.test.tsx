/* @jsxImportSource ../.. */

import type { ComponentInit, SignalifyProps } from '@captainpants/sweeter-core';
import { $val } from '@captainpants/sweeter-core';
import { testRender } from '../../test/testRender.js';

type TestingComponentWithChildrenProps = SignalifyProps<{
    number: number;
    children?: JSX.Element;
    onMount: (num: number) => void;
    onUnMount: (num: number) => void;
}>;

function TestingComponentWithChildren(
    props: TestingComponentWithChildrenProps,
    init: ComponentInit,
): JSX.Element {
    init.onMount(() => {
        $val(props.onMount)($val(props.number));
    });
    init.onUnMount(() => {
        $val(props.onUnMount)($val(props.number));
    });

    return (
        <div>
            <em>{props.number}</em>
            <div>{props.children}</div>
        </div>
    );
}

it('Mount is called in order', () => {
    const mountOrder: number[] = [];
    const unMountOrder: number[] = [];

    const onMount = (num: number) => {
        mountOrder.push(num);
    };
    const onUnMount = (num: number) => {
        unMountOrder.push(num);
    };

    const res = testRender(() => (
        <TestingComponentWithChildren
            number={1}
            onMount={onMount}
            onUnMount={onUnMount}
        >
            <TestingComponentWithChildren
                number={2}
                onMount={onMount}
                onUnMount={onUnMount}
            >
                <TestingComponentWithChildren
                    number={3}
                    onMount={onMount}
                    onUnMount={onUnMount}
                />
            </TestingComponentWithChildren>
            <TestingComponentWithChildren
                number={4}
                onMount={onMount}
                onUnMount={onUnMount}
            />
        </TestingComponentWithChildren>
    ));

    expect(res.nodes).toMatchSnapshot();

    // Depth first order (child before parent)
    expect(mountOrder).toStrictEqual([4, 3, 2, 1]);

    res.dispose();

    // Depth first order (parent before child)
    expect(unMountOrder).toStrictEqual([1, 2, 3, 4]);
});
