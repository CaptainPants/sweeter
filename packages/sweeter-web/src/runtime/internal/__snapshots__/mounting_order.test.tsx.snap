// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount is called in order 1`] = `
NodeList [
  <!--Hooks(TestingComponentWithChildren): onMount, onUnMount-->,
  <div>
    <em>
      1
    </em>
    <div>
      <!--Hooks(TestingComponentWithChildren): onMount, onUnMount-->
      <div>
        <em>
          2
        </em>
        <div>
          <!--Hooks(TestingComponentWithChildren): onMount, onUnMount-->
          <div>
            <em>
              3
            </em>
            <div />
          </div>
        </div>
      </div>
      <!--Hooks(TestingComponentWithChildren): onMount, onUnMount-->
      <div>
        <em>
          4
        </em>
        <div />
      </div>
    </div>
  </div>,
]
`;
